### ISSUE TITLE
[Blazor] Explicitly set values should trump splatted attributes

### ISSUE THREAD

**Issue by mrpmorris** [dotnet/aspnetcore#24233]
> The position of @attributes relative to the position of element attributes is important. When @attributes are splatted on the element, the attributes are processed from right to left (last to first)

https://docs.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-3.1

Attributes having to have a special order is not immediately obvious when trying to understand why an explicitly set parameter value is incorrect.

The position of `@attributes` should be irrelevant. The `@attributes` should be splatted first so that explicitly set parameters cannot be overridden by the consumer of the component.

**Comment by javiercn** [comment]
@mrpmorris thanks for contacting us.

That is less flexible and would make it incredibly hard for components to provide default values for common attributes that are overriden.

While it might have caught you by surprise, this is one of the things that you need to know to use splat attributes effectively.

Also, this would be a big breaking change that we wouldn't likely want to take.

I understand this might not be the behavior that you want, but the current behavior exists for a reason and it is to let developers choose. You can always put your attributes after the splat attributes and they'll always win over them, but with the current behavior, developers have the choice.

**Comment by mrpmorris** [comment]
Hi @javiercn 

To set default values we just need to override SetParametersAsync and set them if null. It's explicit and unambiguous, it is also consistent with how we set default parameters in routes.

It is a breaking change and would need to be done in a major release I expect, but then again dotnet 5 is a major release.

Making the order of html attributes important just doesn't seem like good practice.

**Comment by mrpmorris** [comment]
Update: I see the way it works is consistent with how HTML processes attributes, so that makes sense.

I'll try to find the MS docs for this and suggest an edit to the text, it seems a bit confusing the way it is worded now.

Thanks for your feedback!

**Comment by mrpmorris** [comment]
Update: The priority is opposite to the way HTML works. Take this example

`<div style="color:black" style="color:red">What colour is this text?</div>`

The preceding text is black, because the first style (on the left) takes priority.

Whereas the following

`<h3 attr1="default-value" @attributes=AllOtherParameters attr2="default-value">MyComponent</h3>`

The `attr1` in `AllOtherParameters` overrides the one to the left but not the one to the right.

It's back-to-front, isn't it @javiercn?

**Comment by javiercn** [comment]
@mrpmorris AFAIK there is no order concept on the HTML5 spec.

> To set default values we just need to override SetParametersAsync and set them if null. It's explicit and unambiguous, it is also consistent with how we set default parameters in routes.

This is not an acceptable solution since you give up on all the lifecycle methods at that point.

> It is a breaking change and would need to be done in a major release I expect, but then again dotnet 5 is a major release.

Just because we can make breaking changes it doesn't mean we are willing to make them. For every breaking change we have to  evaluate what the impact is on existing code bases as well as the benefits and drawbacks the change brings. This is not something we are willing to consider for 5.0 and something we are unlikely to consider for future versions.

> The priority is opposite to the way HTML works

AFAIK there is no attribute order concept in HTML, but even if it is, the behavior of attributes in Blazor is not tied to the behavior in HTML. The behavior of `@attributes` is explicit and gives the developer of a component control over how to deal with additional attributes passed in on to the component.

There is no ambiguity on how something will behave based on the comment in the docs. It is true that you need to know how it works, but that's not different than any other alternative and in this case the hint is explicitly specified in the position of the `@attributes` value.

The other two alternatives, which are always overriding explicitly set attributes or never overriding explicitly set values are less powerful than this.

Additionally, the current implementation of `@attributes` performs better than any of the other alternatives since it needs to do less work, as ultimately `@attributes` are just normal attributes applied to a component that behave in the same way as the rest of the attributes in the component.

To summarize my thoughts, I would say that even though I understand how the behavior might surprise you or not be what you expect, we believe that the current behavior as it stands is reasonable and we don't see major reasons to change it in the future given the breaking impact it would have in many customer and component library codebases.

**Comment by mrpmorris** [comment]
> This is not an acceptable solution since you give up on all the lifecycle methods at that point.

I don't understand what this means.

As for the rest, I can understand your reluctance to change it.

**Comment by javiercn** [comment]
> I don't understand what this means.

The way SetParametersAsync is implemented today, if you override you'll have to call base for things like OnInitialized to run when extending componentbase (the most common case). If you override, do something and call base afterwards, it will have no effect since the base implementation will set the parameters. If you call base and then do something afterward, all the OnInititalized and OnParametersSet will run with the wrong set of parameter values. That is something we could fix by introducing an additional lifecycle method that can be overriden, but we haven't seen the need for it yet.

**Comment by mrpmorris** [comment]
Ah no, not that I suppose it matters, but you were thinking of a different approach.

All you have to do is this

```
[Parameter]
public string Something { get; set; }

protected override async SetParametersAsync(ParameterView parameters)
{
  await base.SetParametersAsync(parameters);
  if (Something is null)
    Something = "the default value";
}
```
